EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -545,3 +295,329 @@ void disconnectContactors(uint8_t isContactorError)
+        }
+    }
+
+    switch (toAlsoBeNamed->prevState)
+    {
+        case DONE:
+            contactorError = !toAlsoBeNamed->hasDischargeBeenSet;
+            break;
+
+        case CHARGE_ENABLE_CHECK:
+            contactorError = !(toAlsoBeNamed->hasChargeBeenSet && toAlsoBeNamed->hasDischargeBeenSet);
+            break;
+
+        default: // Something went really wrong in the logic, stop everything
+            disconnectContactors(0);
+            toAlsoBeNamed->state = SHUTDOWN;
+            return;
+    }
+
+    if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
+    {
+        return;
+    }
+
+    auxStatus.contactorError = contactorError;
+
+    if (toAlsoBeNamed->hasDischargeBeenSet && toAlsoBeNamed->hasChargeBeenSet)
+    {
+        auxStatus.startUpSequenceDone = 1;
+        toAlsoBeNamed->state = DONE;
+    }
+
+    osMutexRelease(auxStatus.auxStatusMutex);
+
+    if (!toAlsoBeNamed->hasChargeBeenSet)
+    {
+        toAlsoBeNamed->prevState = toAlsoBeNamed->state;
+        toAlsoBeNamed->state = CHARGE_ENABLE_CHECK;
+        HAL_GPIO_WritePin(CHARGE_ENABLE_GPIO_Port, CHARGE_ENABLE_Pin, GPIO_PIN_SET);
+    }
+    else if (!toAlsoBeNamed->hasDischargeBeenSet)
+    {
+        toAlsoBeNamed->prevState = toAlsoBeNamed->state;
+        toAlsoBeNamed->state = CHARGE_ENABLE_CHECK;
+    }
+}
+
+void done(ToBeNamed* toAlsoBeNamed)
+{
+    const uint8_t commonSense = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+    const uint8_t chargeSense = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+    const uint8_t dischargeSense = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+
+    uint8_t isChargeTurningOn = 0;
+    uint8_t isDischargeTurningOn = 0;
+
+    if (orionStatus.shutOff)
+    {
+        toAlsoBeNamed->state = SHUTDOWN;
+        return;
+    }
+    else if (!commonSense)
+    {
+        disconnectContactors(1);
+        toAlsoBeNamed->state = CONTACTOR_DISCONNECTED;
+    }
+
+    if (canTurnOnChargeContactor(chargeSense))
+    {
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.chargeContactorOverriden = 0;
+        osMutexRelease(orionStatus.orionStatusMutex);
+
+        isChargeTurningOn = 1;
+        toAlsoBeNamed->hasChargeBeenSet = 0;
+        // Turn on charge Contactor
+        HAL_GPIO_WritePin(CHARGE_ENABLE_GPIO_Port, CHARGE_ENABLE_Pin, GPIO_PIN_SET);
+        toAlsoBeNamed->prevState = toAlsoBeNamed->state;
+        toAlsoBeNamed->state = CHARGE_ENABLE_CHECK;
+    }
+    else if (canTurnOnDishargeContactor(dischargeSense))
+    {
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.dischargeContactorOverriden = 0;
+        osMutexRelease(orionStatus.orionStatusMutex);
+
+        isDischargeTurningOn = 1;
+        toAlsoBeNamed->hasDischargeBeenSet = 0;
```

Comment:
@badeniran This might be a future issue, but I have a hunch that instead of having a `isDischargeTurningOn`, we can replace that with a `!hasDischargeBeenSet` usage.

Going further, and this will need testing but I think if we move the checking if contactors disconnected from the end of `blocked` to the beginning, like the other functions, we should be able to simplify the checking logic in `..shouldBeOnButNot` functions

Reply:
I wish we had documented all the things we tried haha, but I think we tried doing the checking logic in the beginning and there were issues. I'll have a look at your first concern though. You might be right, but it may affect `chargeContactorShouldBeOnButNot ` and `dischargeContactorShouldBeOnButNot `

Reply:
Talked with Baba, we'll make a separate issue for this, dw about addressing it here.